libraries/Microsoft.Bot.Builder/Inspection/InspectionSession.cs

```diff
@@ -0,0 +1,83 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license.
+
+using System;
+using System.Net.Http;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Bot.Connector;
+using Microsoft.Bot.Connector.Authentication;
+using Microsoft.Bot.Schema;
+using Microsoft.Extensions.Logging;
+
+namespace Microsoft.Bot.Builder
+{
+    public class InspectionSession
+    {
+        private readonly ConversationReference _conversationReference;
+        private readonly MicrosoftAppCredentials _credentials;
+        private readonly HttpClient _httpClient;
+        private readonly ILogger _logger;
+
+        public InspectionSession(ConversationReference conversationReference, MicrosoftAppCredentials credentials, HttpClient httpClient, ILogger logger)
+        {
+            _conversationReference = conversationReference;
+            _credentials = credentials;
+            _httpClient = httpClient;
+            _logger = logger;
+        }
+
+        public Task<bool> SendAsync(Activity activity, CancellationToken cancellationToken)
+        {
+            activity.ChannelId = _conversationReference.ChannelId;
+            activity.ServiceUrl = _conversationReference.ServiceUrl;
+            if (activity.From == null)
+            {
+                activity.From = _conversationReference.Bot;
+            }
+
+            if (activity.Recipient == null)
+            {
+                activity.Recipient = _conversationReference.User;
+            }
+
+            if (activity.Conversation == null)
+            {
+                activity.Conversation = _conversationReference.Conversation;
+            }
+
+            return SendToConversationAsync(activity, _conversationReference, cancellationToken);
+        }
+
+        protected async Task<bool> SendToConversationAsync(Activity activity, ConversationReference conversationReference, CancellationToken cancellationToken)
+        {
+            if (activity.Timestamp == null)
+            {
+                activity.Timestamp = DateTime.UtcNow;
+            }
+
+            if (activity.ChannelId != conversationReference.ChannelId && activity.RelatesTo == null)
+            {
+                activity.RelatesTo = activity.GetConversationReference();
+            }
+
+            activity.ChannelId = conversationReference.ChannelId;
+            activity.Conversation = conversationReference.Conversation;
```

Comment:
Same question here - should you clone this prior to sending? I'm worried about the external state mutation.

Reply:
it is cloned - right at the beginning